### PR TITLE
Send story context to transform source

### PR DESCRIPTION
**Pull request by NikkiTheBugSlayer**
Issue:

## What I did

This is similar to [#12265 ](https://github.com/storybookjs/storybook/pull/12265), but sends the context to transformSource as third argument so it is not a breaking change.  This is useful for updating the source block when a users changes the controls.

## How to test

I created the button-with-custom-source-code story to demonstrate how context can be used in transformSource.  I also updated enhanceSource to test everything was being sent correctly.

### CODE REVIEW

File: addons/docs/src/blocks/enhanceSource.test.ts [hunk 1]
```diff
@@ -75,9 +94,21 @@ describe('addon-docs enhanceSource', () => {
       expect(enhanceSource(baseContext)).toBeNull();
     });
     it('transformSource', () => {
+      const transformSource = (src?: string) => (src ? `formatted: ${src}` : 'no src');
       const { source } = baseContext.parameters.docs;
       const parameters = { ...baseContext.parameters, docs: { source, transformSource } };
       expect(enhanceSource({ ...baseContext, parameters })).toBeNull();
     });
+    it('receives StoryContext as second argument', () => {
```
Comment (shilman) [inline addons/docs/src/blocks/enhanceSource.test.ts:102]: can we test this with a single test?

Comment (UX-Unicorn) [reply addons/docs/src/blocks/enhanceSource.test.ts]: I removed the extra tests.

**Review by phated — APPROVED** [review]
It's not great that this API will have a different signature than `jsxDecorator`, but I understand if @shilman might think changing the 2nd argument will actually break people.

**Review by shilman — COMMENTED** [review]
Thanks for this contribution @UX-Unicorn! I prefer @phated 's breaking solution -- this API is brand new I don't think enough people are using it yet, so I'm going to make a semver exception and merge that PR instead. Sorry for making you update the tests, I didn't fully understand the PR until I'd read all the related PRs at once.

**Comment by UX-Unicorn** [comment]
> Thanks for this contribution @UX-Unicorn! I prefer @phated 's breaking solution -- this API is brand new I don't think enough people are using it yet, so I'm going to make a semver exception and merge that PR instead. Sorry for making you update the tests, I didn't fully understand the PR until I'd read all the related PRs at once.

@shilman That makes sense.  I just didn't want to have to wait for the breaking change to get merged to get access to the context in transformSource.

**Pull request closed without merging**